src/rules.jl

```diff
@@ -88,8 +87,8 @@ _abs_deriv(x) = signbit(x) ? -one(x) : one(x)
 @define_diffrule Base.log(b, x)    = :( log($x) * inv(-log($b)^2 * $b)                          ), :( inv($x) / log($b)                                                       )
 @define_diffrule Base.ldexp(x, y)  = :( exp2($y)                                                ), :NaN
 
-@define_diffrule Base.mod(x, y)    = :( first(promote(ifelse(isinteger($x / $y), NaN, 1), NaN)) ), :(  z = $x / $y; first(promote(ifelse(isinteger(z), NaN, -floor(z)), NaN)) )
-@define_diffrule Base.rem(x, y)    = :( first(promote(ifelse(isinteger($x / $y), NaN, 1), NaN)) ), :(  z = $x / $y; first(promote(ifelse(isinteger(z), NaN, -trunc(z)), NaN)) )
+@define_diffrule Base.mod(x, y)    = :( z = $x / $y; ifelse(isinteger(z), oftype(z, NaN), one(z)) ), :(  z = $x / $y; ifelse(isinteger(z), oftype(z, NaN), -floor(z)) )
+@define_diffrule Base.rem(x, y)    = :( z = $x / $y; ifelse(isinteger(z), oftype(z, NaN), one(z)) ), :(  z = $x / $y; ifelse(isinteger(z), oftype(z, NaN), -trunc(z)) )
```

Comment:
I'm honestly a bit confused as to why we had the `first(promote` there before. Got any clue @devmotion ?

Reply:
I don't know :shrug: I'm just slightly worried about the variable definitions, I guess `z` might be defined twice in downstream implementations and could possibly leak (IIRC there were some problems even with the tests here at some point and hence I introduced the `let` blocks in the tests). `z` is defined in the current implementation as well, so it seems to be fine. Alternatively, one could use something like `let z = $x / $y; ....; end` but I'm not sure if it's needed.

Reply:
True, I just went with it despite that because the current implementation had `z` in there.. But what do you prefer me doing here? Should we leave as is or introduce `let`?

Reply:
It seems @mcabbott also mentioned it above and considered a gensym alternative (`let` might be easier to debug and inspect). I agree with your comment there and think probably this should be addressed in a separate PR (if we want to change anything). I'm fine with the `z` in this PR.

I noticed another potential issue though: `z` is not necessarily a floating point number, e.g., if `x` and `y` are `Rational`s. Maybe use `float(z)` instead?

Reply:
Ah yes, good catch :+1: I noticed this bug in the bessel functions, etc. but forgot to add it here.